src/HappyEyeBallsConnector.php

```diff
@@ -0,0 +1,37 @@
+<?php
+
+namespace React\Socket;
+
+use React\Dns\Model\Message;
+use React\Dns\Resolver\Resolver;
+use React\EventLoop\LoopInterface;
+use React\EventLoop\TimerInterface;
+use React\Promise;
+use React\Promise\CancellablePromiseInterface;
+use InvalidArgumentException;
+use RuntimeException;
+
+final class HappyEyeBallsConnector implements ConnectorInterface
+{
+    private $loop;
+    private $connector;
+    private $resolver;
+
+    public function __construct(LoopInterface $loop, ConnectorInterface $connector, Resolver $resolver)
+    {
+        $this->loop = $loop;
+        $this->connector = $connector;
+        $this->resolver = $resolver;
+    }
+
+    public function connect($uri)
+    {
+        try {
+            return HappyEyeBallsConnectionBuilder::create($this->loop, $this->connector, $this->resolver, $uri);
```

Comment:
~It's my understanding this piece of logic never throws any exceptions, no?~ I see it will throw an `InvalidArgumentException` for invalid addresses, perhaps move this logic somewhere else to avoid vexing exceptions?

Reply:
Yup updated that and returning a rejected promise instead